feat(react guidelines): initial version

adrianbicanic opened the pull request:
This is an initial version. I guess there are some things that could be unnecessary, as well ass some things that should be added.

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,389 @@
+
+- each component should have it's own folder containing it's logic in a JSX file and it's stylesheet file.
+ (If the app is (or becomes) more complex the components can be placed in additional logically separated sub-folders for better readability and organization).
+- use **camelCase** for general folder naming and **PacalCase** for React components, their stylesheet files and their containing folders 
```
Tpyo: PascalCase

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,389 @@
+- each component should have it's own folder containing it's logic in a JSX file and it's stylesheet file.
+ (If the app is (or becomes) more complex the components can be placed in additional logically separated sub-folders for better readability and organization).
+- use **camelCase** for general folder naming and **PacalCase** for React components, their stylesheet files and their containing folders 
+- each component type's root folder should have an **index.jsx** file which exports all components, so you can later import them from a single place
```
1. We're not using `jsx` extensions, only `js`. When using typescript, we have to use `tsx` (compiler imitation).
2. We're doing this on some projects, but I'm not actually convinced it's a good practice, because it makes code splitting harder/impossible. Might be better to just remove this.

DarkoKukovec replied:
@adrianbicanic

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
````diff
@@ -0,0 +1,389 @@
+```
+
+### Component Declaration
+- don't use React.createClass
````
You should probably start the sentences after a subtitle with a capital letter

adrianbicanic replied:
It's not a sentence, it's a bullet

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
````diff
@@ -0,0 +1,389 @@
+        );
+    }
+}
+```
````
You should also add a TypeScript example

andreicek replied:
Should we separate this in another guide?

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,389 @@
+class ComponentName extends Component {
+    // ...
+    
+    const componentState = {
```
This example doesn't make sense:

- If you're using pure React, you would use `setState` and in this case, you would need to use the `state` property
- This is ok for MobX local state, but you should then mention this

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,389 @@
+
+### Rendering Arrays of Data
+- use array's **map** method to render items from an array
+
```
There should always be a unique `key` property on all created elements

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,389 @@
+    }
+}
+
+// good
```
Looks like you wanted to write something here... perhaps `@action.bound` (not sure it should be here)

adrianbicanic replied:
Removed.

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,389 @@
+    // ...
+    render () {
+        return (
+          <ChildComponent
```
You should add a separate part where you describe how to format components - if it doesn't fit into one row, put every property in a separate row and the closing `>` or `/>` in its own row.

adrianbicanic replied:
Added it.

adrianbicanic commented on JavaScript development/Frameworks/React.md:
````diff
@@ -0,0 +1,389 @@
+        );
+    }
+}
+```
````
Added a TS example.

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
````diff
@@ -0,0 +1,489 @@
+```
+...
+├─ Login
+⎮   └─ Login.jsx
````
This won't work :D

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
````diff
@@ -0,0 +1,489 @@
+}
+```
+
+- if you need to pass more than one CSS class name to a component, use the [`classnames`](https://www.npmjs.com/package/classnames) library
````
Or if you want to use conditional classes...

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
````diff
@@ -0,0 +1,489 @@
+```
+...
+├─ Login
+⎮   └─ Login.js
````
This still won't work :)
You have two `Login.js` files in the same folder

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
````diff
@@ -0,0 +1,489 @@
+```
+
+- use **camelCase** for general folder naming and **PascalCase** for React components, their stylesheet files and their containing folders 
+- each component type's root folder should have an **index.js** file which exports all components, so you can later import them from a single place
````
I would remove this rule. It seems a good idea at first, but it makes code splitting very hard/impossible.

andreicek commented on JavaScript development/Frameworks/React.md:
````diff
@@ -0,0 +1,489 @@
+```
+...
+├─ components
+├─ containers 
````
in some cases this can be pages

adrianbicanic replied:
I commented on that above this example in the guide, and updated it to be more specific.

andreicek commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,489 @@
+});
+    
+// good
+import React, {Component} from 'react';
```
I would go with `React.Component` here. Scales well with TS, and will make sense soon.

andreicek commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,489 @@
+// good
+import React, {Component} from 'react';
+
+class ComponentName extends Component {
```
I prefere `export default class ComponentName`

@DarkoKukovec feel free to chime in.

andreicek commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,489 @@
+// good
+render() {
+  // ...
+  return <ChildComponent />;
```
I like this, but

```
render (
 <ChildComponent />
);
```

scales better. What do you think?

adrianbicanic replied:
Added both.

andreicek commented on JavaScript development/Frameworks/React.md:
````diff
@@ -0,0 +1,489 @@
+};
+```
+
+- you can also declare property types as a **static member** inside a class:
````
this is still not always possible. you need a babel plugin

andreicek commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,489 @@
+        
+        {
+            isConditionMet
+            ? <ConditionallyShownComponent />
```
missing spaces here

andreicek commented on JavaScript development/Frameworks/React.md:
````diff
@@ -0,0 +1,489 @@
+```
+
+
+## Conditional Rendering
````
this isn't a style thing, it has to be this way... does it have sense to leave it in?

andreicek commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,489 @@
+## Rendering Arrays of Data
+- use array's **map** method to render items from an array
+- you should have a unique key for each element
+- **don't use array's index as a key**
```
there is a use case, mention when it's a thing and when it's not. what about local arrays?

andreicek commented on JavaScript development/Frameworks/React.md:
````diff
@@ -0,0 +1,489 @@
+
+``` javascript
+// ...
+import * as styles from 'ComponentName.scss';
````
incorrect path, and why the `*`?

andreicek commented on JavaScript development/Frameworks/React.md:
````diff
@@ -0,0 +1,489 @@
+``` javascript
+import classNames from 'classnames';
+// ...
+import * as styles from 'ComponentName.scss';
````
same as the previous comment.

adrianbicanic commented:
@andreicek @DarkoKukovec I moved the TypeScript guide to it's own folder, and updated React guide.

@andreicek Please merge when everything is confirmed OK.

andreicek commented on JavaScript development/Frameworks/React.md:
````diff
@@ -0,0 +1,536 @@
+If components don't have or change state, use **pure functions**:
+
+``` jsx
+function PureComponent({message}) {
````
add an arrow function example

andreicek commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,536 @@
+
+### Formatting
+
+Use self-closing tags, separated with an empty space, for element that don't have children.
```
for elements that

andreicek commented on JavaScript development/Frameworks/React.md:
````diff
@@ -0,0 +1,536 @@
+<ChildComponent onClick={this.onClickHandler} />
+```
+
+If props don't fit on one line separate them with line breaks and also, put opening and closing tags in separate lines.
````
If there is more than one prop

andreicek commented on JavaScript development/Frameworks/React.md:
````diff
@@ -0,0 +1,536 @@
+}
+```
+
+Omit values for boolean props if default value is **true**.
````
Move this before the functions

andreicek commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,536 @@
+*Checkout [`prop-types`](https://github.com/facebook/prop-types) lib on GitHub for general info.*
+
+In class components:
+- always use prop validation
```
mention typescript here

andreicek commented on JavaScript development/Frameworks/React.md:
````diff
@@ -0,0 +1,536 @@
+
+### Importing React
+``` jsx
+import * as React from 'react';
````
This will work only in TS. Babel is `import React from 'react';`

andreicek commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,536 @@
+
+### Using props
+
+If you are using more than one property in a method, preffer accessing them with **object destructuring**.
```
Mention also extracting complex variables in scope instead of defining them in JSX

```
render() {
    const trasnlatedLabel = label || this.t('translation');
    
    return (
           <Component label={trasnlatedLabel} />
    );
}
```

andreicek commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,536 @@
+  // ..
+};
+
+export default ComponentName;
```
Separate in two code blocks.

andreicek commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,536 @@
+export default PureComponent;
+
+// stateless components with no prop validation
+export default () => (/* ... */);
```
Same

andreicek commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,536 @@
+For iterating arrays:
+- use **map** method to render items
+- you should have a unique key for each element
+- **don't use array's index as a key**
```
Please mention when this is ok. What is book doesn't have an id? What will you do then?

@DarkoKukovec chime in.

DarkoKukovec replied:
This happens if you use indexes: https://twitter.com/ryanflorence/status/920715795088384000
They're only acceptable if you don't have any choice and the list is read-only

andreicek commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,536 @@
+*For general information about MobX, checkout the official documentation:* [`mobx.js.org`](https://mobx.js.org/).
+
+## Decorators
+Use decorators wherever you can because it makes code more readable and declarative.
```
Same as static you may not have support for this.

andreicek commented:
Please update the documentation to forbid default exporting.

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,536 @@
+    <ul>
+      {
+        booksArray.map((book) => {
+          <ListItem key={book.id} title={book.title} /> 
```
There should also be a different example - an example where the items have nested elements, e.g. a link inside of a `<li>`

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,536 @@
+}
+    
+// good
+class ComponentName extends React.Component {
```
Maybe we should think about using something like https://github.com/andreypopp/autobind-decorator
@andreicek

andreicek replied:
I really like this idea. How would do this with `@action.bound`? Would that still be the preferred way of binding actions or would we do this:

```javascript
@autobind
@action
doStuff() { /* code */ }
```

Here is a working example of this: https://codesandbox.io/s/v3jlrpy3w7

I actually think this is better since the methods usually start off not being actions.

andreicek replied:
There is a known problem with `autobind-decorator` and TS. In the issue someone linked developit/decko module that has bind.

Here is an example of this running: https://codesandbox.io/s/ox0p2pym7z

DarkoKukovec replied:
Looks good! Would avoid the `memoize` function trough, seems like it could cause unexpected bugs based on the way it caches the results.

andreicek replied:
Agreed! I would then mention this lib and prefere it instead of `this.fn = this.fn.bind(this);`

@adrianbicanic what do you think?

adrianbicanic replied:
I can comment only regarding how the code looks. Def looks cleaner and more consistent than stuffing 6 methods bindings in the constructor and then binding some more action methods with decorators.

andreicek replied:
OK - let it be so.

- use `developit/decko` lib
- `@bind` over constructor binding
- `@bind` over `@action.bound`
- don't use `memoize`

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,536 @@
+  render() {
+    return (
+      <Heading
+        className={classNames(styles.specialHeading, styles.largeHeading)}
```
Also add an example with conditional classes

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,536 @@
+### Local Component State
+If MobX is used for state management:
+- preffer using MobX *observable* instead of *this.state* for component's local state; don't use both
+- wrap local state in a single object just as it would be in the built in state
```
Make it clear that some name other than `state` should be used. It might also make sense to define a name here, so we're using the same name everywhere. @andreicek

DarkoKukovec commented on JavaScript development/Typescript/TypeScript.md:
````diff
@@ -0,0 +1,117 @@
+
+In stateless components, use regular object destructuring with TypeScript's type annotations.
+``` typescript
+export default ({className?: string, title: string}): JSX.Element => {
````
This is wrong. In this example `string` is not a type, but a variable (default value). Correct example:
`export default ({className, title}: {className?: string, title: string}): JSX.Element => {`

DarkoKukovec commented on JavaScript development/Typescript/TypeScript.md:
````diff
@@ -0,0 +1,117 @@
+}
+```
+
+Use interfaces for props when you have a great amount of props/state to validate or when you have a set of common props used in more than a few components (in which case the interfaces should be imported from a separate file).
````
We should remove the "great amount of props/state to validate" case, and only use interfaces if they're shared across multiple components.

DarkoKukovec commented on JavaScript development/Typescript/TypeScript.md:
````diff
@@ -0,0 +1,117 @@
+React's lifecycle hooks and render method have to be declared public. Local data and methods that are not passed to child components are generally declared private.
+
+``` typescript
+class ComponentName extends Component<IComponentNameProps, IComponentNameState> {
````
Remove interfaces from here, they're not essential for the example, so `{}` should be used.

andreicek commented:
@DarkoKukovec please have a look and merge if ok.

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,598 @@
+export class ComponentName extends React.Component {
+  constructor(args) {
+    super(args);
+    this.onClickHandler = this.onClickHandler.bind(this);
```
Wouldn't say this is necessary bad. This is the recommended way if you don't have decorators available (they're not a standard feature, we're adding them because of mobx)

andreicek replied:
If the project is not using mobx it can still use `bind` decorator so in this case, I would consider this an anti-pattern.

DarkoKukovec replied:
But what if we don't have decorators available?

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,598 @@
+For iterating arrays:
+- use **map** method to render items
+- you should have a unique key for each element
+- **don't use array's index as a key**, unless you have no choice (no unique property in each element of the array)
```
I would also add that the elements need to keep their positions (constant indexes). If not, this causes some very unpredictable bugs.

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
````diff
@@ -0,0 +1,598 @@
+}
+```
+
+Avoid writing any logic inline. Extract it to the outer scope.
````
"outer scope" sounds confusing. Maybe better to change it to something like "outside of the JSX block"

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
````diff
@@ -0,0 +1,598 @@
+/>
+```
+
+Declare functions before passing them as props instead of passing inline anonymous (arrow) functions. They will be needlessy re-rendered each time, thus effecting performance.
````
"re-rendered" is the wrong word here. Functions are not rendered, a better word might be "recreated"

andreicek replied:
"instanced" is even better :)

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,598 @@
+
+For declaring class components:
+- use ES2015 **class syntax**
+- don't use *React.createClass*
```
Maybe add a note:
If you absolutely have to, use the [`create-react-class`](https://www.npmjs.com/package/create-react-class) package instead of `React.createClass` (removed in React 16).

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,598 @@
+  render() {
+    return (
+      <Heading
+        className={classNames(styles.largeHeading, {
```
Should we also follow our rule here of extracting the logic from JSX?

andreicek replied:
eh not sure about this since it's not business logic. your call

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
````diff
@@ -0,0 +1,598 @@
+  }
+}
+```
+
````
Maybe add one more item here defining the max number of lines in a single file? @andreicek

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
````diff
@@ -0,0 +1,598 @@
+```
+
+### Async Await
+In actions implemented as *async* functions, after each *await*, any code that modifies the state should be wrapped in a *runInAction* utility. This is because *async await* is just syntax sugar, and in reallity, a new function is created after each *await*.
````
> This is because *async await* is just syntax sugar, and in reallity, a new function is created after each *await*.

This is only true for the transpiled code, but is also not relevant. The reason is that actions need to be synchronous.

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,598 @@
+  };
+
+  @bind
+  @action
```
The question is should we add an `@action` decorator if it's not needed (like in this case), but the function contains `runInAction`? @andreicek

andreicek replied:
i'm not sure why `action` was used here in the first place.

DarkoKukovec commented on JavaScript development/Typescript/TypeScript.md:
````diff
@@ -0,0 +1,119 @@
+Declare functional components as arrow functions
+
+``` typescript
+export const FunctionalComponent = (/*props*/): JSX.Element => {
````
Is the return type valid in all cases? `null` is a valid use case (when we decide not to render a component), but I'm not sure it's assignable to `JSX.Element`.

andreicek replied:
removed func. components

DarkoKukovec commented on JavaScript development/Typescript/TypeScript.md:
````diff
@@ -0,0 +1,119 @@
+export class ComponentName extends React.Component<CommonProps, {}> {
+  // ...
+}
+```
````
Maybe also add an example when the component needs `CommonProps`, but also has something extra:
```typescript
export class ComponentName extends React.Component<CommonProps & {active?: boolean}, {}> {
  // ...
}
```

DarkoKukovec commented on JavaScript development/Typescript/TypeScript.md:
````diff
@@ -0,0 +1,119 @@
+}
+```
+
+Example with MobX:
````
Is this example relevant? I don't see anything here that wasn't shown in previous examples.

andreicek replied:
yup - removed

DarkoKukovec commented:
Should we also add a Resources section with good tutorials (TS, MobX, etc)?

dingo-d commented on JavaScript development/Frameworks/React.md:
````diff
@@ -0,0 +1,598 @@
+<ChildComponent onClick={this.onClickHandler} />
+```
+
+If there is more than one prop, separate them with line breaks and also, put opening and closing tags in separate lines.
````
Grammar fix: If there is more than one prop, separate them with line breaks, and also put opening and closing tags on separate lines.

dingo-d commented on JavaScript development/Frameworks/React.md:
````diff
@@ -0,0 +1,598 @@
+/>
+```
+
+Declare functions before passing them as props instead of passing inline anonymous (arrow) functions. They will be needlessy re-rendered each time, thus effecting performance.
````
It's _affecting_ performance, not effecting ;)

DarkoKukovec replied:
That was my first thought also, but no:
affecting - touching the emotions; moving.
effecting - cause (something) to happen; bring about.

DarkoKukovec replied:
Actually, I'm not sure... would it be "affecting performance" or "effecting bad performance"?

dingo-d replied:
I think this should be 'affect' because it's a verb, it affects the performance. But 'can have an adverse effect on the performance' could also be used.
Or it could be reformulated to be that 'the effect of passing inline anonymous functions is decreased performance' (sounds bad, and shouldn't be used, but just as an example for when 'effect' can be used :D).

dingo-d commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,598 @@
+*For general information about MobX, checkout the official documentation:* [`mobx.js.org`](https://mobx.js.org/).
+
+### Decorators
+Use decorators wherever you can if your transpiler support's it, because it makes code more readable and declarative.
```
Grammar fix:  ... if your transpiler supports it

dingo-d commented on JavaScript development/Frameworks/React.md:
````diff
@@ -0,0 +1,598 @@
+
+### Actions
+Wrap methods and functions that modify state with *@action*. For binding action methods, use the aforementioned *@bind* decorator from the [`developit/decko`](https://github.com/developit/decko) lib.
+``` jsx
````
Add new line

andreicek commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,536 @@
+
+ComponentName.propTypes = {
+  id: PropTypes.number.isRequired,
+  name: PropTypes.string
```
fixed.

andreicek commented:
you can check it again @DarkoKukovec

DarkoKukovec commented on JavaScript development/Frameworks/React.md:
```diff
@@ -0,0 +1,548 @@
+
+For declaring class components:
+- use ES2015 **class syntax**
+- if you **need** `React.createClass` use [`create-react-class`](https://www.npmjs.com/package/create-react-class) (`createClass` was removed in Reactr 16)
```
`Reactr` typo

DarkoKukovec commented on JavaScript development/Typescript/React.md:
````diff
@@ -0,0 +1,78 @@
+
+In stateless components, use regular object destructuring with TypeScript's type annotations.
+``` typescript
+export const FunctionalComponent = ({classname, title}: {className?: string, title: string}): JSX.Element => {
````
But should we use functional components at all?
